### PR TITLE
Fix uncertainty

### PR DESCRIPTION
**Pull request by mjohnson541**
The pull request solves a number of current issues with uncertainty:

1.  Degeneracies can now be floats
2.  The parser doesn't break when the 'from training reaction' statement comes before and on a different line than 'estimated using template'
3.  Uncertainties can now deal with radical HBI corrections when the saturated species of the radical is not a species in the model

I've added a primitive test that checks that the uncertainties module can load the parameter uncertainty engines (effectively that it doesn't fail to estimate any raw thermo or kinetic uncertainties).  This test should be sufficient that things like 1-3 won't happen again.  

I had to add files to testing_database to pass the tests, but all of the new files are empty.  I had to update the other.py thermo group file because the uncertainty test needed one of the groups in it.

### CODE REVIEW

**Comment by mjohnson541** [comment]
This should fix #1158

**Comment by amarkpayne** [comment]
@mjohnson541, can you fix the code style issues identified by Codacy and then rebase? I should be able to review this by tomorrow evening if so.

**Comment by mjohnson541** [comment]
I've removed the extraneous import, but I'm of the opinion that the rest of the codacy suggested changes for this PR would make harder to read and less clear.

**Comment by mjohnson541** [comment]
Also fixed the iterator one as well now.

File: rmgpy/tools/uncertainty.py [hunk 1]
```diff
@@ -341,37 +326,65 @@ def loadModel(self, chemkinPath, dictionaryPath, transportPath=None):
...
+            if otherSpecies.isIsomorphic(saturatedStruct):
+                return otherSpecies, False
+        else:
+            for key,value in self.database.thermo.libraries.iteritems():
+                for entry in value.entries.values():
+                    if entry.item.isIsomorphic(saturatedStruct):
+                        newSpc = Species().fromAdjacencyList(saturatedStruct.toAdjacencyList())
+                        newSpc.thermo = entry.data
+                        self.speciesList.append(newSpc)
+                        return newSpc, True
```
Comment (amarkpayne) [inline rmgpy/tools/uncertainty.py]: If I understand this correctly, this function is used to return the species object of the saturated molecule in the model to use its sources of thermo estimation to later determine the uncertainty. As you noted, it is possible for a radical species to be present in the model without the corresponding saturated species being present. The idea of your code here is that it is a fall-back for this case, and searches our libraries to see if the molecule exists there.

Specifically, it looks like the uncertainty module loads whichever thermo libraries the user specifies, so as long as the user specifies the same libraries in the same order as they did when they made the RMG model. Assuming the user does this, and we are still unable to find the species in the libraries, could we not assume that the saturated species was estimated by group additivity? If so we should implement a final fall-back method that returns a species object with thermo estimated with group additivity.

tldr: Is a fall-back method for group-additivity needed?

File: rmgpy/tools/uncertainty.py [hunk 2]
```diff
@@ -341,37 +326,65 @@ def loadModel(self, chemkinPath, dictionaryPath, transportPath=None):
...
     def extractSourcesFromModel(self):
         """
         Extract the source data from the model using its comments.
         Must be done after loading model and database to work.
         """
         self.speciesSourcesDict = {}
+        ignoreSpcs = []
         for species in self.speciesList:
-            source = self.database.thermo.extractSourceFromComments(species)
+            if not species in ignoreSpcs:
```
Comment (amarkpayne) [inline rmgpy/tools/uncertainty.py]: Is this line necessary? I think the only way for a species to be added to `ignoreSpcs` is for it to have gone through this block once, and since we only loop over each specie in `self.speciesList` once this line should always be true.

File: rmgpy/tools/uncertainty.py [hunk 3]
```diff
@@ -341,37 +326,65 @@ def loadModel(self, chemkinPath, dictionaryPath, transportPath=None):
...
+        """
+        
+        molecule = species.molecule[0]
+        assert molecule.isRadical(), "Method only valid for radicals."
+        saturatedStruct = molecule.copy(deep=True)
+        saturatedStruct.saturate()
+        for otherSpecies in self.speciesList:
+            if otherSpecies.isIsomorphic(saturatedStruct):
+                return otherSpecies, False
+        else:
```
[the remark this answers is not in the document]
Comment (mliu49) [reply rmgpy/tools/uncertainty.py]: Codacy is flagging it because you don't need the else. It either returns or continues after the for loop. The same thing would be accomplished by removing the else and reducing the indentation of everything after.

File: rmgpy/tools/uncertainty.py (hunk 1 above)
Comment (mliu49) [reply rmgpy/tools/uncertainty.py]: In `extractSourceFromComments`, HBI and GAV are considered a single source. Retrieving the saturated structure is only done if the thermo was from two sources, ie. HBI on top of a library value or HBI on top of a QM value.

File: rmgpy/tools/uncertainty.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (amarkpayne) [reply rmgpy/tools/uncertainty.py]: That's fair. I do agree with @mjohnson541 though, I like having the else there so that it is clear that the code only runs in the event that the for loop does not succeed. Would you object to changing the else statement with a comment like `# else if no species were found, check for the species in the thermo libraries`

File: rmgpy/tools/uncertainty.py [hunk 4]
```diff
@@ -341,37 +326,65 @@ def loadModel(self, chemkinPath, dictionaryPath, transportPath=None):
...
+            for key,value in self.database.thermo.libraries.iteritems():
```
[the remark this answers is not in the document]
Comment (mliu49) [reply rmgpy/tools/uncertainty.py]: It might be better to create the species first, then call `self.database.thermo.getThermoDataFromLibraries()`.

As an additional note, the species can be directly using `newSpc = Species(molecule=[saturatedStruct])` instead of conversion via adjlist.

File: rmgpy/tools/uncertainty.py (hunk 1 above)
Comment (amarkpayne) [reply rmgpy/tools/uncertainty.py]: Gotcha, so a fall-back to GAV is not only unnecessary but would give the wrong result (since we only would get here if GAV was not used)?

**Comment by mliu49** [comment]
Sorry for the late comment, but I just remembered that I made a quick fix for the saturated species problem a long time ago. You can see it [here](https://github.com/mliu49/RMG-Py/commit/2d08944d863aa8215364cd134d7028bea2fc892b), but I think your changes are pretty similar and possibly more robust.

File: rmgpy/tools/uncertainty.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mliu49) [reply rmgpy/tools/uncertainty.py]: I think the cleanest way might be to return at the very end of the method. Returning early is also fine, but I would prefer not having the else. A comment would be helpful though.
```
for species in self.speciesList:
    if ... :
        result = something
        break
else:
    result = something_else

return result
```

File: rmgpy/tools/uncertainty.py (hunk 1 above)
Comment (mliu49) [reply rmgpy/tools/uncertainty.py]: I think so, if I understood correctly.

File: rmgpy/tools/uncertainty.py (hunk 2 above)
Comment (mjohnson541) [reply rmgpy/tools/uncertainty.py]: So ignoreSpcs are saturated species that are needed for getting thermo uncertainty correct, but aren't in the model, and they get appended to the end of speciesList so it can find them later.  However we do not want the sources for that molecule indexed in the same way species that are actually in the model are to avoid confusion later in processing.

File: rmgpy/data/kinetics/family.py [hunk 5]
```diff
@@ -2140,8 +2140,11 @@ def assignWeightsToEntries(entryNestedList, weightedEntries, N = 1):
             
             # Discard the last line, unless it's the only line!
             # The last line is 'Estimated using ... for rate rule (originalTemplate)'
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/data/kinetics/family.py]: added

File: rmgpy/tools/uncertainty.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/tools/uncertainty.py]: This should be much cleaner now

File: rmgpy/tools/uncertainty.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/tools/uncertainty.py]: I've removed the else

**Comment by mjohnson541** [comment]
I've made all of the changes and I added a small fix for some group names that end in parentheses that solves an issue I ran into.

File: rmgpy/data/kinetics/family.py [hunk 6]
```diff
@@ -2140,8 +2140,12 @@ def assignWeightsToEntries(entryNestedList, weightedEntries, N = 1):
...
-            if len(lines) == 1:
-                comment = lines[0]
+            #if from training reaction is in the first line append it to the end of the second line and skip the first line
+            if not 'Average of' in kinetics.comment:
+                if 'from training reaction' in lines[0]:
+                    comment = lines[1]+' '+lines[0]
```
Comment (mliu49) [inline rmgpy/data/kinetics/family.py]: Do we need to do this, or could we just get rid of the line with 'from training reaction' since it doesn't seem like we use it?

Comment (mjohnson541) [reply rmgpy/data/kinetics/family.py]: It gets looked for in retrieveOriginalEntry, which expects it to be on the same line as the primary comment.

File: rmgpy/tools/uncertaintyTest.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        uncertainty.compileAllSources()
+        uncertainty.assignParameterUncertainties()
+        
+        thermoUnc = uncertainty.thermoInputUncertainties
+        kineticUnc = uncertainty.kineticInputUncertainties
+        
+        logging.error(thermoUnc)
+        logging.error(kineticUnc)
+        
+        self.assertEquals(thermoUnc,[1.5, 1.5, 2.0, 1.8999999999999999, 3.1000000000000001, 1.5, 1.8999999999999999, 2.0, 2.0, 1.8999999999999999, 2.2000000000000002, 1.8999999999999999, 2.0, 1.5])
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/tools/uncertaintyTest.py]: Good idea!

File: rmgpy/tools/uncertaintyTest.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+            
+        uncertainty.extractSourcesFromModel()
+        uncertainty.compileAllSources()
+        uncertainty.assignParameterUncertainties()
+        
+        thermoUnc = uncertainty.thermoInputUncertainties
+        kineticUnc = uncertainty.kineticInputUncertainties
+        
+        logging.error(thermoUnc)
+        logging.error(kineticUnc)
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/tools/uncertaintyTest.py]: I thought it might be useful to someone trying to debug if the test fails, but I don't have a strong opinion about it.  I'll remove it.

File: rmgpy/tools/uncertainty.py [hunk 9]
```diff
@@ -341,37 +326,66 @@ def loadModel(self, chemkinPath, dictionaryPath, transportPath=None):
                                                               dictionaryPath=dictionaryPath,
                                                               transportPath=transportPath)
 
+    def retrieveSaturatedSpeciesFromList(self,species):
+        """
+        Given a radical `species`, this function retrieves the saturated species objects from a list of species objects.
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/tools/uncertainty.py]: added

File: rmgpy/tools/uncertainty.py [hunk 10]
```diff
@@ -341,37 +326,66 @@ def loadModel(self, chemkinPath, dictionaryPath, transportPath=None):
...
+        saturatedStruct.saturate()
+        for otherSpecies in self.speciesList:
+            if otherSpecies.isIsomorphic(saturatedStruct):
+                return otherSpecies, False
+        
+        #couldn't find saturated species in the model, try libraries
+        newSpc = Species(molecule=[saturatedStruct])
+        thermo = self.database.thermo.getThermoDataFromLibraries(newSpc)
+        
+        if not thermo is None:
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/tools/uncertainty.py]: done

File: rmgpy/data/kinetics/family.py (hunk 6 above)
Comment (mliu49) [reply rmgpy/data/kinetics/family.py]: I don't think retrieveOriginalEntry uses it. If you look at line 2150 below, it splits the comment and takes the fourth token, which is the original template which was matched, which is the only argument given to retrieveOriginalEntry.

Comment (mjohnson541) [reply rmgpy/data/kinetics/family.py]: Oh, you're right it grabs that information from the comment on the rule not the original kinetics comment even though it's present in both.

Comment (mjohnson541) [reply rmgpy/data/kinetics/family.py]: This should be fixed now

File: rmgpy/data/kinetics/family.py [hunk 11]
```diff
@@ -2162,7 +2166,9 @@ def assignWeightsToEntries(entryNestedList, weightedEntries, N = 1):
                                     re.sub(r"(?<=]) \+ (?!Average)", ",'",      # + sign between average and non-average
                                     re.sub(r"(?<!]) \+ (?!Average)", "','",     # + sign between non-averages
                                     comment)))))))
-
+            
+                evalCommentString = evalCommentString.replace("),",")',")
```
Comment (mliu49) [inline rmgpy/data/kinetics/family.py]: Do you have an example of the type of comment that this deals with? Would it be possible to update the regex above instead of adding this line?

Comment (mjohnson541) [reply rmgpy/data/kinetics/family.py]: I can't remember where I got the original failures, but this 'made up' comment replicates the parsing problem.  I just added parentheses around Csrad everywhere it showed up.  I think I missed one, but it shouldn't matter in the parsing.  

"""Average of [Average of [O2b;O_(Csrad)] + Average of [O_atom_triplet;O_(Csrad) + CH2_triplet;O_(Csrad)] + Average of [H_rad;O_(Csrad) + Average of [Ct_rad/Ct;O_(Csrad) from training reaction 0] + Average of [O_pri_rad;O_(Csrad) + Average of [O_rad/NonDeC;O_(Csrad) + O_rad/NonDeO;O_(Csrad)]] + Average of [Cd_pri_rad;O_(Csrad)] + Average of [CO_pri_rad;O_(Csrad)] + Average of [C_methyl;O_(Csrad) + Average of [C_rad/H2/Cs;O_Csrad + C_rad/H2/Cd;O_(Csrad) + C_rad/H2/O;O_(Csrad)] + Average of [C_rad/H/NonDeC;O_(Csrad)] + Average of [Average of [C_rad/Cs3;O_(Csrad)]]]]]
Estimated using template [Y_rad_birad_trirad_quadrad;O_(Csrad)] for rate rule [CH_quartet;O_(Csrad)]
Euclidian distance = 2.0
family: Disproportionation
"""

After processing it becomes:  
"[['O2b;O_(Csrad)'],['O_atom_triplet;O_(Csrad)','CH2_triplet;O_(Csrad)'],['H_rad;O_(Csrad),['Ct_rad/Ct;O_(Csrad) from training reaction 0'],['O_pri_rad;O_(Csrad),['O_rad/NonDeC;O_(Csrad)','O_rad/NonDeO;O_(Csrad)']],['Cd_pri_rad;O_(Csrad)'],['CO_pri_rad;O_(Csrad)'],['C_methyl;O_(Csrad),['C_rad/H2/Cs;O_Csrad','C_rad/H2/Cd;O_(Csrad)','C_rad/H2/O;O_(Csrad)'],['C_rad/H/NonDeC;O_(Csrad)'],[['C_rad/Cs3;O_(Csrad)']]]]]"

This can't be evaluated because where there should be "),'" there are "),".  It should be possible to adapt the regex, but I think it's a lot harder to do and harder to read than my fix.

**Comment by mliu49** [comment]
I added some stuff to the unit test you wrote, and I fixed the regex to incorporate the bug you discovered. Let me know your thoughts.

A test that checks Cantera output files is failing for an unknown reason, which I'm working on figuring out right now. I can't reproduce the failure locally.

**Comment by mjohnson541** [comment]
Your changes look good to me.

**Comment by mliu49** [comment]
I rebased and removed the debugging commits. If @amarkpayne could take a final look (since I also made some changes), I think this is ready to be merged once the tests pass.

**Comment by mjohnson541** [comment]
@mliu49 did you want to merge this in?

**Comment by mliu49** [comment]
Sure, I was hoping for @amarkpayne to do so, but I can do it now.

**Comment by mjohnson541** [comment]
@amarkpayne doesn't have merge permissions

**Comment by mliu49** [comment]
Ah, sorry, I didn't realize that.
